Draft - Liberty Deep Dive blog (#2273)
OpenLiberty/blogs: posts/2022-05-28-liberty-deep-dive.adoc

---
layout: post
title: "Introducing a hands-on technical deep dive on Liberty!"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/ryan-storey
author_github: https://github.com/ryan-storey
seo-title: Introducing a hands-on technical deep dive on Liberty! - OpenLiberty.io

> seo-description: The Liberty Deep Dive will teach you how to build a RESTful microservice on Liberty with Jakarta EE and MicroProfile, document your REST API, persist data to a database, secure your applications using JSON Web Tokens (JWT), test your applications and more.

Reviewer: ```suggestion
seo-description: Introducing an end-to-end tutorial on cloud-native Java application development and deployment using Jakarta EE, MicroProfile and Liberty. The tutorial covers building RESTful microservice, documenting your REST API, externalizing configuration, persisting data to a database, securing application using JSON Web Tokens (JWT), adding health checks, providing metrics, testing using Testcontainers, building container images, and deploying to Kubernetes using Operator.
```

> blog_description: "TThe Liberty Deep Dive will teach you how to build a RESTful microservice on Liberty with Jakarta EE and MicroProfile, document your REST API, persist data to a database, secure your applications using JSON Web Tokens (JWT), test your applications and more."

Reviewer: ```suggestion
blog_description: "Introducing an end-to-end tutorial on cloud-native Java application development and deployment using Jakarta EE, MicroProfile and Liberty"
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Introducing a hands-on technical deep dive on Liberty!
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> We are excited to announce the release of our newest tutorial, A Technical Deep Dive on Liberty! Our team have been hard at work to produce a single tutorial which covers many cloud-native concepts and bundles them into one application. This tutorial is aimed to help developers, from Liberty novices to those more experienced, develop, build, test, and deploy cloud-native Java applications and microservices on Open Liberty. You will learn how to build a RESTful microservice on Liberty with Jakarta EE and MicroProfile, documenting your REST API, persisting data to a database, securing your applications using JSON Web Tokens (JWT), testing your applications and more. You will use Maven throughout this exercise to build the microservice and to interact with the running Liberty instance. Lastly, you’ll build a container image for the microservice and deploy it to Kubernetes using the Open Liberty Operator.

Reviewer: ```suggestion
We are excited to announce the release of our newest tutorial, A Technical Deep Dive on Liberty!

Our team has been hard at work to produce a single end-to-end tutorial that covers important cloud-native concepts, developing and deploying applications and microservices using open cloud-native Java APIs in Jakarta EE and MicroProfile while following best practices for working with Liberty. This tutorial is aimed to help developers, from Liberty novices to those more experienced, develop, build, test, and deploy your applications on Liberty running in Kubernetes and clouds.

In this tutorial, you will learn how to build RESTful microservice on Liberty with Jakarta EE and MicroProfile, documenting your REST API, externalizing configuration, persisting data to a database, securing your application using JSON Web Tokens (JWT), adding health checks, providing metrics, testing your applications and more. You will also look at building container image for your Java microservice and deploying it to Kubernetes using the Open Liberty Operator.
```

Reviewer 2: ```suggestion
We are excited to announce the release of our newest tutorial, A Technical Deep Dive on Liberty!

Our team has been hard at work to produce a single end-to-end tutorial that covers developing and deploying applications and microservices by using open cloud-native Java APIs in Jakarta EE and MicroProfile while following best practices for working with Liberty. This tutorial is aimed to help developers, from Liberty novices to those more experienced, develop, build, test, and deploy your applications on Liberty running in Kubernetes and clouds.

In this tutorial, you will learn how to:
- build a RESTful microservice on Liberty with Jakarta EE and MicroProfile 
- document your REST API 
- externalize your configuration
- persist data to a database 
- secure your application using JSON Web Tokens (JWT) 
- add health checks 
- provide metrics 
- test your applications,  and more. 

You will also build a container image for your Java microservice and deploy it to Kubernetes by using the Open Liberty Operator.
```

Reviewer 3: While I appreciate that we _are_ excited to get this published after all our hard work, I think we're centering ourselves a bit much here. I doubt the reader cares about our excitement and hard work so much as what the tutorial will do for them.

Keep the "We are excited..." sentence if you like, but lose the exclamation mark/point (!).

Then start the next paragraph with "A Technical Deep Dive on Liberty is an end-to-end tutorial that walks you through how to develop and deploy cloud-native Java applications in Jakarta EE and MicroProfile. The tutorial is aimed at Java developers, from Liberty novices to those more experienced with writing applications to run on Liberty."

Then change the lead-in sentence to: "In A Technical Deep Dive on Liberty, you will learn how to:" (then have the bulleted list like David suggests as it's nice and easy to skim down.

Shouldn't that last sentence also be bullets rather than a separate sentence? So:

- build a container image for your Java microservice
- deploy the container image to Kubernetes using the Open Liberty Operator

You can access the tutorial directly link:{url-prefix}/guides/liberty-deep-dive.html[through this link].

> Excitingly, this lab is also available for you to try in our cloud-hosted environment, without having to worry about local setup and prerequisites. Simply head over to the link:{url-prefix}/guides/liberty-deep-dive.html[tutorial page], and click the green *'Run in cloud'* button. Alternatively, you can head over to the workshop via link:https://ibm.biz/LibertyDeepDive[https://ibm.biz/LibertyDeepDive].

Reviewer: ```suggestion
Conveniently, this lab is also available for you to try in our cloud-hosted environment, without having to worry about local setup and prerequisites. Simply head over to the link:{url-prefix}/guides/liberty-deep-dive.html[tutorial page] and click the green *'Run in cloud'* button. Alternatively, you can access the cloud-hosted tutorial via link:https://ibm.biz/LibertyDeepDive[https://ibm.biz/LibertyDeepDive] directly.
```

Reviewer 3: Which approach do we want them to take? If we're tracking clicks (as the ibm.biz link implies), we should provide _only_ the direct link. If they end up in the normal guides page, they can work out for themselves what the green button is for.

Reviewer 3: How about:

"If you don't want to worry about local setup and prerequisites, try link:https://ibm.biz/LibertyDeepDive[A Technical Deep Dive on Liberty in our cloud-hosted environment].

Alternatively, if you prefer, you can complete link:{url-prefix}/guides/liberty-deep-dive.html[A Technical Deep Dive on Liberty on your own laptop]."

> Excitingly, this lab is also available for you to try in our cloud-hosted environment, without having to worry about local setup and prerequisites. Simply head over to the link:{url-prefix}/guides/liberty-deep-dive.html[tutorial page], and click the green *'Run in cloud'* button. Alternatively, you can head over to the workshop via link:https://ibm.biz/LibertyDeepDive[https://ibm.biz/LibertyDeepDive].

Reviewer 2: ```suggestion
Conveniently, this lab is also available for you to try in our cloud-hosted environment, without having to worry about local setup and prerequisites. Simply head over to the link:{url-prefix}/guides/liberty-deep-dive.html[tutorial page] and click the green *'Run in cloud'* button. Alternatively, you can access the cloud-hosted tutorial directly at link:https://ibm.biz/LibertyDeepDive[https://ibm.biz/LibertyDeepDive].
```

image::/img/blog/deepdive.png[screen capture of the liberty deep dive UI,width=70%,align="center"]

> == Your feedback is welcomed

Reviewer 2: ```suggestion
== We welcome your feedback
```

We should change this in the template as well

> Tell us about your experience with our guides by responding to the questions in the *Where to next?* section at the end of the tutorial. We’d love to hear from you!

Reviewer: ```suggestion
Tell us about your experience with this Technical Deep Dive on Liberty by responding to the questions in the *Where to next?* section at the end of the tutorial. We’d love to hear from you!
```

Reviewer: Can we include a screen capture of the feedback section if not provide a direct link to the corresponding GH issue for users to file issues?

Reviewer 3: I think we need to be consistent on the title of the tutorial. So:

"Tell us about your experience with A Technical Deep Dive on Liberty by responding to the questions in the *Where to next?* section at the end of the tutorial. We’d love to hear from you!"

author_github: https://github.com/ryan-storey
seo-title: Introducing a hands-on technical deep dive on Liberty! - OpenLiberty.io
seo-description: Introducing an end-to-end tutorial on cloud-native Java application development and deployment using Jakarta EE, MicroProfile and Liberty. The tutorial covers building RESTful microservice, documenting your REST API, externalizing configuration, persisting data to a database, securing application using JSON Web Tokens (JWT), adding health checks, providing metrics, testing using Testcontainers, building container images, and deploying to Kubernetes using Operator.
blog_description: "Introducing an end-to-end tutorial on cloud-native Java application development and deployment using Jakarta EE, MicroProfile and Liberty"
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Introducing a hands-on technical deep dive on Liberty!
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

We are excited to announce the release of our newest tutorial, A Technical Deep Dive on Liberty.

> A Technical Deep Dive on Liberty is an end-to-end tutorial that walks you through how to develop and deploy cloud-native Java applications in Jakarta EE and MicroProfile. The tutorial is aimed at Java developers, from Liberty novices to those more experienced with writing applications to run on Liberty. This tutorial is aimed to help developers, from Liberty novices to those more experienced, develop, build, test, and deploy your applications on Liberty running in Kubernetes and clouds.

Reviewer 3: Remove the sentence "This tutorial is aimed to help developers, from Liberty novices to those more experienced, develop, build, test, and deploy your applications on Liberty running in Kubernetes and clouds." It just duplicates what's already said and coming next anyway.

Reviewer: This is from the tutorial's abstract.  I am thinking it is redundant so good to drop it.

In A Technical Deep Dive on Liberty, you will learn how to:

* build a RESTful microservice on Liberty with Jakarta EE and MicroProfile 
* document your REST API 
* externalize your configuration
* persist data to a database 
* secure your application using JSON Web Tokens (JWT) 
* add health checks 
* provide metrics 
* build a container image for your Java microservice
* deploy the container image to Kubernetes using the Open Liberty Operator

> * test your applications, and more. 

Reviewer 3: A suggestion (@yeekangc - what do you think? depends on whether you want to publish this week more)...could the benefits of some of these be explained a little?
eg "easily document your REST API so that other developers can use it"; "externalize your configuration so that it can be easily, and automatically, be changed throughout the development lifecycle"; "add health checks to automatically verify whether your microservice is available or not"; "provide metrics to monitor the activity of your microservice".

Can probably drop the ", and more." bit from the last bullet. There's quite a lot listed already!

Reviewer: Good point!  Great if we can do this and it will be helpful to the readers too.

Reviewer: - easily document your REST API for developers to consume your microservice
- externalize your configuration to make your application portable and easier to update
- add health checks to easily report the statuses of your microservice (for orchestration decision)
- provide metrics to monitor activity and performance of your application
- test your applications easily with containers (for true-to-production testing)

Reviewer 3: Thought of these too but see what @yeekangc thinks (not sure some are perfectly accurate):

* "build a RESTful microservice on Liberty with Jakarta EE and MicroProfile and make it available for requests from other applications"
* "persist data to a database so that your microservice can look up or store information from requests"
* "secure your application using JSON Web Tokens (JTW) so that only authenticated and authorized applications can request information from your microservice"

Reviewer: I feel we don't need the additional qualifications for "build a RESTful microservice on Liberty with Jakarta EE and MicroProfile" and ""persist data to a database" since they are or should be obvious to a person familiar with the art?

The one for JWT is good though.
